### PR TITLE
fix(runtime): string scanners must not read past exact-sized slice allocations (#6085)

### LINKED ISSUE #6085: Runtime split()/parseFloat() read past exact-sized slice allocations -> intermittent AV (c0000005) on hot paths
## Symptom

Access violation (c0000005) with empty stderr — the compiled game dies silently or freezes (last presented frame stays on screen, input appears dead). Faulting reads land just past the end of a heap page (`0x...FFF8`-style addresses). The crash is layout-sensitive: a relink can hide or resurface it, which makes it look unreproducible.

## Trigger

Running a string through `split()` + `parseFloat()` on a hot path (every frame). Perry's runtime string scanners appear to read one word past the end of their own exact-sized slice allocations; with enough fresh allocations per second, one eventually lands flush against an unmapped page and faults.

We hit this parsing a `label|1.23|4.56\n...`-style blob returned from a native FFI every frame: **6/6 crashes within 7-29 seconds**, reproduced across two different link layouts. One-shot parses carry the same risk per call, just with much lower odds.

## Minimal repro sketch

1. Any FFI (or plain TS source) producing a multi-line `1.23|4.56\n` blob.
2. Per frame: `for (const line of blob.split('\n')) { const parts = line.split('|'); parseFloat(parts[0]); ... }`
3. On Windows x64 this AVs within a minute on an affected build.

## Workaround we shipped

Replaced the delimited-text ABI with a numeric one (one FFI call per scalar, f64 across the boundary; strings cross whole and are only drawn, never parsed). Also tail-padded engine-allocated FFI strings by 16 zero bytes — note this cannot protect Perry-internal slice allocations, which is why the ABI change was required.

## Environment

- perry 0.5.1208 (also reproduced on the 0.5.11xx line)
- Windows 11 Pro 26200, x64, lld-link
- Crash addresses consistently at end-of-heap-page boundaries; minidumps available on request

### PR DESCRIPTION
**Pull request by proggeramlug**
## Root cause

Perry heap-string payloads are **exact-sized**: `string_storage_alloc` reserves `size_of::<StringHeader>() + byte_len` bytes — no NUL terminator, no tail padding — and the bytes are **not guaranteed to be valid UTF-8**. `Buffer.toString()`, FFI blobs, and WTF-8 lone-surrogate strings all reach `js_string_from_bytes` with raw bytes, and `string_as_str` hands those bytes to `str::from_utf8_unchecked`.

Several scanners then decoded that view with std's **UTF-8-validity-assuming** iterators (`str::chars()`, `str::encode_utf16()`). `core::str::validations::next_code_point` reads the continuation bytes of a multi-byte lead with `unwrap_unchecked` — no bounds check. So a payload whose **last sequence is a truncated multi-byte lead** reads 1–3 bytes **past the end of its own allocation**.

That is exactly the reported shape: reads just past the end of a heap page, layout-sensitive (a relink hides or resurfaces it), fatal only when an allocation happens to land flush against an unmapped page (`0x...FFF8`-style faulting addresses).

**How the repro reaches it.** A native FFI / `Buffer` blob is not required to be well-formed UTF-8. Splitting such a blob on an ASCII delimiter (`\n`, `|`) can cut *inside* a multi-byte sequence, and each `split()` part is a fresh **exact-sized** allocation ending in a dangling lead byte. Any subsequent scan of that part over-reads. This matches the reporter's "6/6 crashes within 7–29 seconds" on a per-frame `split()` + parse hot path, where the churn of fresh exact-sized allocations eventually places one against a guard page.

### OOB sites fixed (each proven to fault, see below)

| Site | Read class |
|---|---|
| `string/mod.rs` `utf16_offset_to_byte_offset` / `byte_offset_to_utf16_index` — reached by `slice` / `substring` / `substr` / `startsWith(pos)` / `endsWith(pos)` | `chars()` / `encode_utf16()` continuation-byte read past `byte_len` |
| `string/char_ops.rs` `js_string_char_code_at`, `js_string_code_point_at` | `chars()` walk past `byte_len` |
| `string/char_ops.rs` `js_string_at` | `encode_utf16().collect()` over the whole payload |
| `string/char_ops.rs` `js_string_to_char_array` (spread `[...s]`) | `chars()` walk past `byte_len` |
| `string/slice_ops.rs` `trim` / `trimStart` / `trimEnd` | `str::trim_matches` → `chars()` |
| `string/slice_ops.rs` `toLowerCase` / `toUpperCase` | `str::to_lowercase/to_uppercase` → `chars()` |
| `string/split.rs` empty-delimiter `split("")` | `chars()` walk past `byte_len` |

### Explicitly NOT a fix I am claiming

- **`parseFloat` is already bounds-safe.** `parse_float_bytes` / `float_prefix_end` / `trim_leading_js_whitespace` guard every index with `i < n` and operate on the exact byte slice. I audited it and changed nothing there. The fault is on the **scan** side, not the parse side.
- **`split()`'s non-empty-delimiter path is already bounds-safe** (byte search + exact-length copies). Only its `split("")` character walk over-read.
- `js_string_char_at` and `js_string_index_of` were already bounds-driven; they get guard-page tests as regression cover, not fixes.
- #6255 previously bounded `decode_wtf8_units` (`padStart`/`padEnd`). This PR closes the remaining sites in the same class.

## Convention enforced: bounds-driven scanning

Perry strings carry an explicit `byte_len` and **no terminator**, and the payload is not guaranteed well-formed. Adding a NUL terminator would mean touching every allocation site, growing every string by a byte, and would still not make `chars()` safe (it stops at a *lead* byte, not a NUL). So the invariant enforced is: **never read past `byte_len`; never assume UTF-8 validity.**

All the walks above now go through one shared helper, `string::wtf8_step`, which classifies a sequence from its lead byte exactly as `compute_utf16_len_wtf8` already does (so cursor walks agree with the `utf16_len` in the header) and reads continuation bytes **only** via bounds-checked `get()`, substituting 0 for bytes that do not exist. Well-formed input decodes byte-for-byte identically.

Case conversion additionally now round-trips lone surrogates and truncated sequences **verbatim** instead of depending on `from_utf8_unchecked`.

### 1. Guard-page tests (deterministic, red → green)

New `crates/perry-runtime/src/string/tests_guard_page.rs` (`#[cfg(all(test, unix))]`): `mmap` two pages, `mprotect(PROT_NONE)` the second, and place the `StringHeader` so the **last payload byte is the last byte of the first page**. Any read past `byte_len` hits the guard page. Payload is `C3 A9 41 C3` ("é", "A", dangling 2-byte lead) with `utf16_len (3) != byte_len (4)` so the non-ASCII decode path is actually taken.

Against the **pre-fix** scanners, release build:

```
char_code_at_does_not_read_past_payload              (signal: 10, SIGBUS)
code_point_at_does_not_read_past_payload             (signal: 10, SIGBUS)
string_at_does_not_read_past_payload                 (signal: 10, SIGBUS)
to_char_array_does_not_read_past_payload             (signal: 10, SIGBUS)
trim_and_case_conversion_do_not_read_past_payload    (signal: 10, SIGBUS)
split_by_empty_delimiter_does_not_read_past_payload  (signal: 10, SIGBUS)
slice_and_substring_do_not_read_past_payload         (signal: 10, SIGBUS)
```

(macOS maps the guard-page read to SIGBUS; it is the same access violation Windows reports as `c0000005`. In a debug build the same reads instead panic inside `core/src/str/validations.rs:48` — literally `next_code_point`'s `unwrap_unchecked` — which pinpoints the over-read.)

After the fix, all 12 pass:

```
test result: ok. 12 passed; 0 failed
```

### 2. Behavior parity

`cargo test -p perry-runtime -- --test-threads=1`: **1244 passed, 0 failed.** (The suite has pre-existing order-dependent flakes when run in parallel — a different unrelated test fails on each parallel run, on main as well as here.)

TS repro compiled with this build — per-frame `blob.split('\n')` → `line.split('|')` → `parseFloat`, 500 frames, plus edge cases (`""`, `"1.23e5"`, `"  1.5"`, `"abc"`, `"Infinity"`, `"-Infinity"`, `"-0"` (incl. `1/x` sign check), `"3."`, `".5"`, `"1e"`, `"+2.5"`, `"0x10"`, `"5px"`, empty parts, trailing newline) and every scanner this PR touches over multi-byte text (`split`, spread, `charCodeAt`, `codePointAt`, `at`, `charAt`, `slice`, `substring`, `indexOf`, `trim*`, case conversion, `.length`):

- **vs a stock `main` build: output byte-identical** → this fix changes no observable behavior.
- **vs `node --experimental-strip-types`: identical**, except one line — `"😀x".at(0) === "\ud83d"` — which is the pre-existing, documented lone-surrogate/WTF-8 categorical gap (`at()` substitutes U+FFFD). Verified it prints the same wrong value on unmodified `main`, so it is untouched by this PR and out of scope.

`cargo fmt --all -- --check` clean.

Fixes #6085

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved JavaScript string character/index/code point access for non-ASCII, truncated multi-byte, and lone-surrogate data.
  * Updated splitting, trimming, and case conversion to avoid out-of-bounds reads on malformed or truncated UTF-8-like payloads.
  * Preserved correct UTF-16/astral and surrogate-pair behavior for valid Unicode inputs.
* **Tests**
  * Added guarded-memory regression tests to verify safe decoding and correct results for both valid and truncated sequences.

### CODE REVIEW

File: crates/perry-runtime/src/string/split.rs [hunk 1]
```diff
@@ -129,15 +129,26 @@ pub extern "C" fn js_string_split_n(
...
+        let src = unsafe { slice::from_raw_parts(string_data(s), (*s).byte_len as usize) };
+        let mut parts: Vec<*mut StringHeader> = Vec::new();
+        let mut i = 0usize;
+        while i < src.len() {
+            let (advance, _, _) = crate::string::wtf8_step(src, i);
+            let end = (i + advance).min(src.len());
+            let seq = &src[i..end];
+            parts.push(js_string_from_bytes(seq.as_ptr(), seq.len() as u32));
+            i = end;
+        }
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/string/split.rs]: Fixed in 3525a31dd — and the hazard was real, so thanks.

**Provenance:** this is **pre-existing on `main`**, not introduced or widened by this PR. `main`'s empty-delimiter path has the identical shape (`str_data.chars().map(js_string_from_bytes).collect::<Vec<*mut StringHeader>>()`, then `js_array_alloc` + fill afterwards); my diff only changed how the sequences are *produced*. Fixing it anyway since it's in a function this PR touches.

**What I changed (both paths):**
- Root the source string (`root_string_ptr`) and the result array (`root_raw_mut_ptr`) in a `RuntimeHandleScope`, and re-read both after every allocation — handles are scanned *and* rewritten on C4b evacuation, a `Vec<*mut …>` is neither.
- Publish each part into the rooted array **as it is built**, so from that moment the array keeps it alive.
- The non-empty-delimiter path had the same hazard **plus** the #5062 dangling-source class: it copied from `&str` slices borrowed from the source payload *after* `string_storage_alloc` had already run. Parts are now carried as byte **offsets** (valid across a move) and the source address is re-read from the handle before each copy.
- Empty-delimiter sequences are staged through a 4-byte stack buffer before allocating, since `js_string_from_bytes` allocates first and copies second.
- `length` is now bumped **incrementally**. `js_array_alloc` does not zero its element storage, so publishing the full length up front (as the old code did — safe only because its fill loop never allocated) would let a mid-loop GC scan uninitialized slots as JSValues. My rewrite allocates inside the fill loop, so this mattered.

**On proof — being straight with you:** I could **not** get this to fire deterministically. A split-heavy repro (400 rounds × 60-part splits with concurrent allocation churn) under `PERRY_GC_FORCE_EVACUATE=1 PERRY_GC_VERIFY_EVACUATION=1` reports `corrupted 0` on the **pre-fix stock `main` binary** across repeated runs. So I'm claiming this as a fix-by-inspection of a genuine unsoundness (raw heap pointers are not roots and are not rewritten), not as a reproduced crash. The same repro passes on the fixed build, and output is byte-identical to Node.

`cargo test -p perry-runtime -- --test-threads=1`: 1245 passed, 0 failed.

File: crates/perry-runtime/src/string/slice_ops.rs [hunk 2]
```diff
@@ -212,60 +212,153 @@ pub(crate) fn is_js_whitespace(c: char) -> bool {
...
+            if !is_ws {
+                last_non_ws_end = next;
+            }
+            i = next;
+        }
+        end = last_non_ws_end;
+    }
+
+    (start, end.max(start))
+}
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/string/slice_ops.rs:270]: Fixed in 3525a31dd. Good catch — this one **is** a real bug my diff introduced, and my own doc comment asserted the opposite ("a truncated/invalid tail sequence is never whitespace"), which was simply false.

The mechanism is exactly as you describe: `wtf8_step` zero-fills missing continuation bytes, so a dangling `E2 80` decodes as `cp = 0x2000` (EN QUAD) — which **is** in the JS `WhiteSpace` set — and `trim`/`trimEnd` would then silently **delete** those bytes. That's data loss, and it also diverged from `case_convert`, which already gated on a complete sequence (`advance == end - i`) and passed truncated tails through verbatim.

Fix: new `js_whitespace_seq_at` helper gates the whitespace test on the sequence being **complete** — `i + advance <= bytes.len()` — before consulting `is_js_whitespace`. A truncated tail is therefore non-whitespace and is preserved byte-for-byte, consistent with `case_convert`.

Regression test added (`trim_preserves_truncated_multibyte_tail`), asserting both directions:
- payload `41 41 E2 80` (a truncated `E2 80`) survives `trim` / `trimEnd` **byte-for-byte**;
- payload `41 E2 80 80` (a **complete** U+2000) is still trimmed to `41`.

Both run against the `PROT_NONE` guard page, so the scan is also proven not to over-read while doing it.

File: crates/perry-runtime/src/string/split.rs [hunk 3]
```diff
@@ -129,66 +129,139 @@ pub extern "C" fn js_string_split_n(
...
             let (sh, data_ptr) = string_storage_alloc(byte_len);
+            let s_now = s_handle.get_raw_const_ptr::<StringHeader>();
+            let part_ptr = string_data(s_now).add(offset);
             let utf16_len = if src_is_ascii {
                 byte_len
             } else {
-                compute_utf16_len(part.as_ptr(), byte_len)
+                compute_utf16_len(part_ptr, byte_len)
             };
             init_string_header(sh, utf16_len, byte_len, byte_len, 0, 0);
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/string/split.rs:247]: Confirmed and fixed in 19a7e2274. This one is a genuine bug, and it lands squarely inside this PR's own threat model — a payload that is *not* well-formed UTF-8 is the entire reason #6085 exists.

I verified it rather than taking it on inspection. Instrumented `js_string_split_n` on the branch with your exact input `[0x80, '|', 0xF0]`:

```
SOURCE byte_len=3 utf16_len=3 is_ascii_string=true    <- the aggregate lies
  part[0] bytes=[80] byte_len=1 utf16_len(RECORDED)=1 utf16_len(CORRECT)=0
  part[1] bytes=[F0] byte_len=1 utf16_len(RECORDED)=1 utf16_len(CORRECT)=2
```

Exactly as you predicted: a stray continuation byte is **0** UTF-16 units and a 4-byte lead is **2**, but the fast path stamped **1** on both. That wrong `.length` then propagates into every downstream index/length operation.

The flags half is real too, and slightly worse than stated. With `"\uD800|B"` (`ED A0 80 | B`):

```
SRC flags=1 utf16_len=3 byte_len=5
  PART[0] bytes=[ED, A0, 80] flags=0  (isWellFormed=true)   <- wrong
```

Worth flagging one nuance: `js_string_from_bytes` **also** hardcodes `flags = 0` (it only derives `utf16_len`), so literally "mirroring its flag derivation" would still drop the surrogate. The flag has to be derived from the part's **own bytes**, which is what I did.

**Fix:**
- Replaced the aggregate `is_ascii_string(s)` gate with an actual **byte scan of the source payload**. A genuinely all-ASCII source provably has all-ASCII parts, so that shortcut *is* sound per-part — I kept it (it's the common case and the hot path), but it's now gated on something true rather than on an aggregate that malformed bytes can satisfy.
- Otherwise `utf16_len` is computed per part via the bounded `compute_utf16_len` over that part's own bytes.
- `STRING_FLAG_HAS_LONE_SURROGATES` is derived per part via a new bounds-driven `bytes_have_lone_surrogate()` helper, on **both** the non-empty and empty-delimiter paths. A part carries the flag only if the surrogate actually landed in it (the `"B"` part correctly stays unflagged).

**Tests added:** `split_parts_get_metadata_from_their_own_bytes` (asserts 0 and 2, and asserts the precondition that `is_ascii_string` really does misfire — so the test would catch a regression to the aggregate check) and `split_parts_preserve_lone_surrogate_flag` (flagged part → `isWellFormed() == false`; clean part → unflagged; covers the `split("")` path too).

`cargo test -p perry-runtime -- --test-threads=1`: 1247 passed, 0 failed. Guard-page suite green. Node parity repro: byte-identical to a stock `main` build, and identical to Node.

**Comment by proggeramlug** [comment]
Heads up: this merged at `3525a31dd`, which is one commit **before** the fix for the third review finding (split parts inheriting wrong `utf16_len`/`flags` from the aggregate `is_ascii_string` check). That fix is therefore **not on `main`** — it is now up on its own as #6295.
